internal/controllers/import_controller_v3.go

```diff
@@ -0,0 +1,356 @@
+		return reconcileDelete(ctx, capiCluster)
+	}
+
+	return r.reconcileNormal(ctx, capiCluster, rancherCluster)
+}
+
+func (r *CAPIImportManagementV3Reconciler) reconcileNormal(ctx context.Context, capiCluster *clusterv1.Cluster,
+	rancherCluster *managementv3.Cluster,
+) (ctrl.Result, error) {
+	log := log.FromContext(ctx)
+
+	err := r.RancherClient.Get(ctx, client.ObjectKeyFromObject(rancherCluster), rancherCluster)
+	if apierrors.IsNotFound(err) {
+		shouldImport, err := util.ShouldAutoImport(ctx, log, r.Client, capiCluster, importLabelName)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+		if !shouldImport {
+			log.Info("not auto importing cluster as namespace or cluster isn't marked auto import")
+			return ctrl.Result{}, nil
+		}
+
+		if err := r.RancherClient.Create(ctx, &managementv3.Cluster{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace:    capiCluster.Namespace,
+				GenerateName: "c-",
+				Labels: map[string]string{
+					capiClusterOwner:          capiCluster.Name,
+					capiClusterOwnerNamespace: capiCluster.Namespace,
+					ownedLabelName:            "",
+				},
+			},
+			Spec: managementv3.ClusterSpec{
+				DisplayName: capiCluster.Name,
+				Description: "CAPI cluster imported to Rancher",
+			},
+		}); err != nil {
+			return ctrl.Result{}, fmt.Errorf("error creating rancher cluster: %w", err)
+		}
+
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	if err != nil {
+		log.Error(err, fmt.Sprintf("Unable to fetch rancher cluster %s", client.ObjectKeyFromObject(rancherCluster)))
+
+		return ctrl.Result{}, err
+	}
+
+	if managementv3.ClusterConditionAgentDeployed.IsTrue(rancherCluster) {
+		log.Info("agent already deployed, no action needed")
+		return ctrl.Result{}, nil
+	}
+
+	// get the registration manifest
+	manifest, err := getClusterRegistrationManifest(ctx, rancherCluster.Name, rancherCluster.Name, r.RancherClient, r.InsecureSkipVerify)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if manifest == "" {
+		log.Info("Import manifest URL not set yet, requeue")
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	log.Info("Creating import manifest")
+
+	remoteClient, err := r.remoteClientGetter(ctx, capiCluster.Name, r.Client, client.ObjectKeyFromObject(capiCluster))
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("getting remote cluster client: %w", err)
+	}
+
+	if err := createImportManifest(ctx, remoteClient, strings.NewReader(manifest)); err != nil {
+		return ctrl.Result{}, fmt.Errorf("creating import manifest: %w", err)
+	}
+
+	log.Info("Successfully applied import manifest")
+
+	return ctrl.Result{}, nil
+}
+
+func (r *CAPIImportManagementV3Reconciler) rancherClusterToCapiCluster(ctx context.Context, clusterPredicate predicate.Funcs) handler.MapFunc {
+	log := log.FromContext(ctx)
+
+	return func(_ context.Context, o client.Object) []ctrl.Request {
+		labels := o.GetLabels()
+		capiCluster := &clusterv1.Cluster{ObjectMeta: metav1.ObjectMeta{
+			Name:      labels[capiClusterOwner],
```

Comment:
Should we do an explicit check for the label first?

Reply:
It now checks explicitly for the two required labels.